HDDS-6129. OM has thread unsafe issues that can cause OM shutdown.

captainzmc opened the pull request:
## What changes were proposed in this pull request?

See: https://issues.apache.org/jira/browse/HDDS-6129

## How was this patch tested?

Existing UT

adoroszlai commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/RepeatedOmKeyInfo.java:
```diff
@@ -66,7 +68,7 @@ public static RepeatedOmKeyInfo getFromProto(RepeatedKeyInfo
    */
   public RepeatedKeyInfo getProto(boolean compact, int clientVersion) {
-    List<KeyInfo> list = new ArrayList<>();
+    List<KeyInfo> list = new CopyOnWriteArrayList<>();
     for(OmKeyInfo k : omKeyInfoList) {
```
Since the concurrent modification is reported during iteration, I think the problem is with `omKeyInfoList`, not `list`.

captainzmc replied:
Thanks @adoroszlai  can review this.
This exception is 100% reproducible when we use FIO concurrent testing.  Looking at the stack, we can see that it was a thread safety problem of RepeatedOmKeyInfo, and the problem did not reappear after this fix.

I checked that a similar error had occurred in OM before, when TreeMap was used in OmMultipartKeyInfo：https://issues.apache.org/jira/browse/HDDS-2356

Recreating this problem with UT can be complex, and currently we only encounter this problem with FUSE and FIO concurrent writes. We have not found this problem in other operations.

adoroszlai replied:
Repro:

```java
  @Test
  public void threadSafety() throws InterruptedException {
    final OmKeyInfo key = getKeyInfo(1);
    final RepeatedOmKeyInfo subject = new RepeatedOmKeyInfo(key);
    final RepeatedOmKeyInfoCodec codec = new RepeatedOmKeyInfoCodec(true);
    final AtomicBoolean failed = new AtomicBoolean();
    ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true)
        .build();
    threadFactory.newThread(() -> {
      for (int i = 0; i < 1000000; i++) {
        try {
          codec.toPersistedFormat(subject);
        } catch (Exception e) {
          e.printStackTrace();
          failed.set(true);
        }
      }
    }).start();
    threadFactory.newThread(() -> {
      for (int i = 0; i < 10000; i++) {
        subject.addOmKeyInfo(key);
      }
    }).start();
    final long start = Time.monotonicNow();
    while (!failed.get() && (Time.monotonicNow() - start < 5000)) {
      Thread.sleep(100);
    }
    assertFalse(failed.get());
  }
```

```
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
	at java.util.ArrayList$Itr.next(ArrayList.java:859)
	at org.apache.hadoop.ozone.om.helpers.RepeatedOmKeyInfo.getProto(RepeatedOmKeyInfo.java:70)
	at org.apache.hadoop.ozone.om.codec.RepeatedOmKeyInfoCodec.toPersistedFormat(RepeatedOmKeyInfoCodec.java:50)
```

bharatviswa504 replied:
I think the correct way to fix is implement copyObject in RepeatedOmKeyInfo properly right now it is returning same object. I believe this might be exposed due to recent fixes, where we are adding RepeatedOmKeyInfo to cache to avoid cleanup when override.

For example refer similar thing is observed in OMKeyInfo. 
https://issues.apache.org/jira/browse/HDDS-2322

Some background when we get from Table#get API's if we get new object, then there should be no race conditions or multiple threads modifiying same object and we will not see this issue.
You can refer this where we introduced CopyObject to solve this kind of issue.

https://issues.apache.org/jira/browse/HDDS-2344

captainzmc replied:
Thanks @adoroszlai and @bharatviswa504 for you comment. I will modify this PR as suggested.

adoroszlai reviewed (changes requested):
Thanks @captainzmc for working on this.

I'm not sure the right list was changed to thread-safe one (see inline comment).

Also, I don't see how existing unit tests would cover this, since they have been passing with and without the change.

captainzmc commented:
I have modified this PR by referring to the OmMultipartKeyInfo modification in HDDS-2322. Also added UT. Can you help take antoher look? @bharatviswa504 @adoroszlai

bharatviswa504 reviewed (commented):
+1 LGTM.
Now with this change, not sure the added test makes sense. But i am fine with it.

adoroszlai commented:
> Now with this change, not sure the added test makes sense.

Agreed, I posted the test method only as a simple way to repro the problem.

captainzmc commented:
> > Now with this change, not sure the added test makes sense.
> 
> Agreed, I posted the test method only as a simple way to repro the problem.

Tnanks @adoroszlai and @bharatviswa504 for the review.
I changed the original test, where I passed [subject.copyObject()](https://github.com/apache/ozone/pull/2938/files#diff-e56566deb75e0f977c634733076a0f049a58cf21186c3e6e5b4d83808bb4951aR139). This will get an error before this patch,  after we re-new the object there will be no errors here.

captainzmc commented:
Merged this. Thanks @adoroszlai and @bharatviswa504 for the review.
